File: 프로그래머스/lv1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    rotated_board = list(map(list, zip(*board[::-1])))
    result = 0
    stack = []
    
    for b in rotated_board:
        while b and b[-1] == 0:
            b.pop()
    
    for move in moves:
        if rotated_board[move - 1]:
            doll = rotated_board[move - 1].pop()
            
            if stack and doll == stack[-1]:
                stack.pop()
                result += 2
            else:
                stack.append(doll)
    
    return result
            

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int n = board.length;
        int[] tops = new int[n];
        Deque<Integer> basket = new ArrayDeque<>();
        int count = 0;

        // Initialize tops array
        for (int i = 0; i < n; i++) {
            tops[i] = getTopItem(board, i);
        }

        for (int move : moves) {
            int col = move - 1;
            if (tops[col] == n) continue; // Empty column

            int item = board[tops[col]++][col];
            
            if (!basket.isEmpty() && basket.peek() == item) {
                basket.pop();
                count += 2;
            } else {
                basket.push(item);
            }
        }

        return count;
    }

    private int getTopItem(int[][] board, int col) {
        for (int i = 0; i < board.length; i++) {
            if (board[i][col] != 0) {
                return i;
            }
        }
        return board.length;
    }
}